{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "162c625e-4102-4520-a7d1-7b69eaec8a6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tqdm.auto import tqdm\n",
    "\n",
    "import tfscreen\n",
    "\n",
    "from tfscreen import build_condition_dataframes\n",
    "from tfscreen import generate_libraries\n",
    "from tfscreen import generate_phenotypes\n",
    "from tfscreen import transform_and_mix\n",
    "from tfscreen import initialize_population\n",
    "from tfscreen import simulate_growth\n",
    "from tfscreen import sequence_conditions\n",
    "\n",
    "import random\n",
    "import copy\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8ac75438-4453-4aa9-844c-da889a830eab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# -----------------------------------------------------------------------------\n",
    "# Ensemble information\n",
    "\n",
    "# spreadsheet defining ensemble and interactions with iptg\n",
    "ensemble_spreadsheet = \"input/ensemble.xlsx\"\n",
    "\n",
    "# Spreadsheet with affects of all mutations on each state in ensemble\n",
    "ddG_spreadsheet = \"input/ddG.xlsx\"\n",
    "\n",
    "T = 310      # K\n",
    "R = 0.001987 # kcal/mol/K\n",
    "\n",
    "# This sets how the thermodynamic variable maps proportionally to output. For\n",
    "# the lac repressor, this is 12 because we dervied the thermodynamics \n",
    "# parameters under conditions where there was a 12-fold excess of the lac \n",
    "# repressor to operator. Fully bound lac repressor thus corresponds to a \n",
    "# an observable value of 1/12 in the operator-bound state--hence 12 here.  \n",
    "scale_obs_by = 12\n",
    "\n",
    "# -----------------------------------------------------------------------------\n",
    "# Library definition information\n",
    "\n",
    "# wildtype amino acid sequence (single letter, uppercase)\n",
    "aa_sequence = \\\n",
    "\"\"\"\n",
    "VKPVTLYDVAEYAGVSYQTVSRVVNQASHVSAKTREKVEAAMAELNYIPNRVAQQLAGK\n",
    "QSLLIGVATSSLALHAPSQIVAAIKSRADQLGASVVVSMVERSGVEACKAAVHNLLAQR\n",
    "VSGLIINYPLDDQDAIAVEAACTNVPALFLDVSDQTPINSIIFSHEDGTRLGVEHLVAL\n",
    "GHQQIALLAGPLSSVSARLRLAGWHKYLTRNQIQPIAEREGDWSAMSGFQQTMQMLNEG\n",
    "IVPTAMLVANDQMALGAMRAITESGLRVGADISVVGYDDTEDSSCYIPPLTTIKQDFRL\n",
    "LGQTSVDRLLQLSQGQAVKGNQLLPVSLVKRKTTLAPNTQTASP\n",
    "\"\"\"\n",
    "\n",
    "# sites that are mutated. sites should either be wildtype \n",
    "# amino acids or non-uppercase letters. non-uppercase letters\n",
    "# are interpreted as sub-library identifiers. \n",
    "mutated_sites = \\\n",
    "\"\"\"\n",
    "VKPVTLYDVAEYAGVSYQTVSRVVNQASH111111111111111111111111111111\n",
    "QS2222222222222222222222222222222222222VERSGVEACKAAVHNLLAQR\n",
    "VSGLIINYPLDDQDAIAVEAACTNVPALFLDVSDQTPINSIIFSHEDGTRLGVEHLVAL\n",
    "GHQQIALLAGPLSSVSARLRLAGWHKYLTRNQIQPIAEREGDWSAMSGFQQTMQMLNEG\n",
    "IVPTAMLVANDQMALGAMRAITESGLRVGADISVVGYDDTEDSSCYIPPLTTIKQDFRL\n",
    "LGQTSVDRLLQLSQGQAVKGNQLLPVSLVKRKTTLAPNTQTASP\n",
    "\"\"\"\n",
    "\n",
    "# Number of first letter in amino sequence. Should match \n",
    "# numbering in ddG spreadsheet\n",
    "seq_starts_at = 1\n",
    "\n",
    "# Should we make internal combinations (within O1, O2, etc.)\n",
    "internal_doubles = False \n",
    "\n",
    "# highest order combination to make (1 -- alone, 2 -- pairs,\n",
    "# 3 -- triplicates, etc.)\n",
    "max_num_combos = 2\n",
    "\n",
    "# What codon to place at mutated sites\n",
    "degen_codon = \"nnt\"\n",
    "\n",
    "# -----------------------------------------------------------------------------\n",
    "# Transformation and library composition\n",
    "\n",
    "# Define how many transformants to include for each sub-library. Keys are \n",
    "# tuples of strings. ('1',) means O1; ('2',) means O2; ('1','2') means O1/O2. \n",
    "# if you set internal_doubles to True above, you can then add ('1','1') and \n",
    "# ('2','2') to get the doubles within each pool. \n",
    "\n",
    "# sizes: how many clones will you get for the smallest transformation of this\n",
    "# library\n",
    "transform_sizes = {('1',):1e5,\n",
    "                   ('2',):1e5,\n",
    "                   ('1','2'):1e6}\n",
    "\n",
    "# mixture: what is the mixture ratio of the library. This works together with \n",
    "# sizes. For the example values, the final mixed library would have 1e5 * 1\n",
    "# O1, 1e5 * 1, O2, and and 1e6 * 10 O1/O2. Note that the **ratio** matters, \n",
    "# not the absolute value. You do not need to have all 1E10 clones... in fact,\n",
    "# keep as small as possible while maintinaing ratios. I'd recommend setting \n",
    "# the lowest count library bit to 1, then scaling rest appropriately.\n",
    "library_mixture = {('1',):1,\n",
    "                   ('2',):1,\n",
    "                   ('1','2'):10}\n",
    "\n",
    "# The number of plasmids per cell is set by sampling from a Poisson \n",
    "# distribuation. If 0 or None, each cell gets exactly one plasmid. Otherwise,\n",
    "# do Poisson sampling.\n",
    "lambda_value = 3\n",
    "\n",
    "# Maximum number of plasmids per cell. The lower this number, the faster the\n",
    "# analysis; however, if you make it too small relative to lambda_value, you'll\n",
    "# distort the Poisson distribution by cutting off the long right tail. 10 should\n",
    "# be safe for lambda ~ 2-3. \n",
    "max_num_plasmids = 10 \n",
    "\n",
    "# Standard deviation on the base growth rate vis-a-vis wildtype. \n",
    "mut_growth_rate_std = 0.01\n",
    "\n",
    "\n",
    "# -----------------------------------------------------------------------------\n",
    "# Pre selection culture set up\n",
    "\n",
    "# Number of cells that come out of glycerol stock\n",
    "# (NIH says 40%)\n",
    "num_thawed_colonies = 1e7\n",
    "\n",
    "# Overnight cell culture volume\n",
    "overnight_volume_in_mL = 10\n",
    "\n",
    "# Grow cells to this cfu/mL before diluting into IPTG \n",
    "# (OD600 of 0.35 on spec is 89370229 cfu/mL)\n",
    "pre_iptg_cfu_mL = 90000000\n",
    "\n",
    "# Grow out this long before putting into selection conditions\n",
    "iptg_out_growth_time = 30\n",
    "\n",
    "# Dilute by this factor when adding to IPTG tubes\n",
    "post_iptg_dilution_factor = 0.2/10.2\n",
    "\n",
    "\n",
    "\n",
    "# -----------------------------------------------------------------------------\n",
    "# Selection conditions\n",
    "\n",
    "# condition_blocks is a list of dictionaries defining different growt\n",
    "# conditions. Each dictionary should have a \"marker\" key with a value of either\n",
    "# \"kanR\" or \"pheS\", a \"select\" key with a value of either 0 or 1, an \"iptg\"\n",
    "# key with a list of iptg concentrations (in mM), and a \"time\" key with a list\n",
    "# of sample times (in minutes)\n",
    "\n",
    "condition_blocks = [{\"iptg\":[0,1],\n",
    "                     \"marker\":\"kanR\",\n",
    "                     \"select\":0,\n",
    "                     \"time\":[80,95,110]},\n",
    "                    {\"iptg\":[0,1],\n",
    "                     \"marker\":\"pheS\",\n",
    "                     \"select\":0,\n",
    "                     \"time\":[80,95,110]},\n",
    "                    {\"iptg\":[0,0.0001,0.001,0.003],\n",
    "                     \"marker\":\"kanR\",\n",
    "                     \"select\":1,\n",
    "                     \"time\":[160,180,200]},\n",
    "                    {\"iptg\":[0.01,0.03,0.1,1.0],\n",
    "                     \"marker\":\"kanR\",\n",
    "                     \"select\":1,\n",
    "                     \"time\":[145,165,180]},\n",
    "                    {\"iptg\":[0, 0.0001, 0.001, 0.003, 0.01, 0.03, 0.1, 1.0],\n",
    "                     \"marker\":\"pheS\",\n",
    "                     \"select\":1,\n",
    "                     \"time\":[95,110,125]}]\n",
    "                     \n",
    "# -----------------------------------------------------------------------------\n",
    "# Data collection parameters\n",
    "\n",
    "total_num_reads = 1e9\n",
    "\n",
    "# This populates the replicate column in the output dataframes, exactly as \n",
    "# happens experimentally. Not really used, but could be helpful if you wanted \n",
    "# to combine multiple sims. \n",
    "replicate = 1\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c65aa88b-c892-4e62-974b-548ce7b959f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "calculating phenotypes\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6cabd839f3c04041aedfc26f29b7d685",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/224890 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "storing phenotypes\n",
      "Generating initial populations\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2b599c671f6440458d97e482e996ac92",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/10200000 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Getting bacterial growth rates\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "505dcded7ae24e0b9069b45643dd0e45",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1102233 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Growing populations in specified conditions\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5f9779aed0914d259c1f687c9cc8d9c5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/20 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequencing final populations\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5388bcd6875b483ab0100a268123f471",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/60 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building final dataframe\n"
     ]
    }
   ],
   "source": [
    "# -----------------------------------------------------------------------------\n",
    "# Build library\n",
    "\n",
    "libraries, genotype_df = generate_libraries(aa_sequence=aa_sequence,\n",
    "                                            mutated_sites=mutated_sites,\n",
    "                                            seq_starts_at=seq_starts_at,\n",
    "                                            max_num_combos=max_num_combos,\n",
    "                                            internal_doubles=internal_doubles,\n",
    "                                            degen_codon=degen_codon)\n",
    "\n",
    "condition_df, condition_df_with_time = build_condition_dataframes(condition_blocks,\n",
    "                                                                  replicate=replicate)\n",
    "\n",
    "# -----------------------------------------------------------------------------\n",
    "# Calculate phenotypes\n",
    "\n",
    "# Calculate phenotype. Each row is indexed by a unique genotype\n",
    "phenotype_df, genotype_df = generate_phenotypes(genotype_df=genotype_df,\n",
    "                                                condition_df=condition_df,\n",
    "                                                ensemble_spreadsheet=ensemble_spreadsheet,\n",
    "                                                ddG_spreadsheet=ddG_spreadsheet,\n",
    "                                                scale_obs_by=scale_obs_by,\n",
    "                                                mut_growth_rate_std=mut_growth_rate_std,\n",
    "                                                T=T,\n",
    "                                                R=R)\n",
    "\n",
    "# -----------------------------------------------------------------------------\n",
    "# Sample from library and grow out\n",
    "\n",
    "# Sample from the main library\n",
    "input_library = transform_and_mix(libraries=libraries,\n",
    "                                  transform_sizes=transform_sizes,\n",
    "                                  library_mixture=library_mixture,\n",
    "                                  max_num_plasmids=max_num_plasmids,\n",
    "                                  lambda_value=lambda_value)\n",
    "\n",
    "\n",
    "# Create initial populations and growth rates for each cell under all relevant \n",
    "# conditions\n",
    "init_output = initialize_population(input_library=input_library,\n",
    "                                    phenotype_df=phenotype_df,\n",
    "                                    genotype_df=genotype_df,\n",
    "                                    condition_df=condition_df,\n",
    "                                    num_thawed_colonies=num_thawed_colonies,\n",
    "                                    overnight_volume_in_mL=overnight_volume_in_mL,\n",
    "                                    pre_iptg_cfu_mL=pre_iptg_cfu_mL,\n",
    "                                    iptg_out_growth_time=iptg_out_growth_time,\n",
    "                                    post_iptg_dilution_factor=post_iptg_dilution_factor)\n",
    "    \n",
    "bacteria, ln_pop_array, bact_condition_k = init_output\n",
    "\n",
    "# Simulate the growth if of each condition. This creates a dictionary keying \n",
    "# conditions to ln_pop vectors for each bacterium. condition_df_with_time\n",
    "# gets updated to include the total population at each \n",
    "condition_pops = simulate_growth(ln_pop_array,\n",
    "                                 bact_condition_k,\n",
    "                                 condition_df,                                          \n",
    "                                 condition_df_with_time)\n",
    "\n",
    "count_df, condition_df_with_time = sequence_conditions(condition_pops,\n",
    "                                                       condition_df_with_time,\n",
    "                                                       genotype_df,\n",
    "                                                       bacteria,\n",
    "                                                       total_num_reads)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0a59dd8b-5e76-4050-b288-3ddf02f62c9a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genotype</th>\n",
       "      <th>genotype_number</th>\n",
       "      <th>condition</th>\n",
       "      <th>base_condition</th>\n",
       "      <th>base_condition_number</th>\n",
       "      <th>time</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-0-0.0-80</td>\n",
       "      <td>1-kanR-0-0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>80</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-0-0.0-95</td>\n",
       "      <td>1-kanR-0-0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>95</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-0-0.0-110</td>\n",
       "      <td>1-kanR-0-0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>110</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-0-1.0-80</td>\n",
       "      <td>1-kanR-0-1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>80</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-0-1.0-95</td>\n",
       "      <td>1-kanR-0-1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>95</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-0-1.0-110</td>\n",
       "      <td>1-kanR-0-1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>110</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.0-160</td>\n",
       "      <td>1-kanR-1-0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>160</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.0-180</td>\n",
       "      <td>1-kanR-1-0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.0-200</td>\n",
       "      <td>1-kanR-1-0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.0001-160</td>\n",
       "      <td>1-kanR-1-0.0001</td>\n",
       "      <td>3</td>\n",
       "      <td>160</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.0001-180</td>\n",
       "      <td>1-kanR-1-0.0001</td>\n",
       "      <td>3</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.0001-200</td>\n",
       "      <td>1-kanR-1-0.0001</td>\n",
       "      <td>3</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.001-160</td>\n",
       "      <td>1-kanR-1-0.001</td>\n",
       "      <td>4</td>\n",
       "      <td>160</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.001-180</td>\n",
       "      <td>1-kanR-1-0.001</td>\n",
       "      <td>4</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.001-200</td>\n",
       "      <td>1-kanR-1-0.001</td>\n",
       "      <td>4</td>\n",
       "      <td>200</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.003-160</td>\n",
       "      <td>1-kanR-1-0.003</td>\n",
       "      <td>5</td>\n",
       "      <td>160</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.003-180</td>\n",
       "      <td>1-kanR-1-0.003</td>\n",
       "      <td>5</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.003-200</td>\n",
       "      <td>1-kanR-1-0.003</td>\n",
       "      <td>5</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.01-145</td>\n",
       "      <td>1-kanR-1-0.01</td>\n",
       "      <td>6</td>\n",
       "      <td>145</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.01-165</td>\n",
       "      <td>1-kanR-1-0.01</td>\n",
       "      <td>6</td>\n",
       "      <td>165</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.01-180</td>\n",
       "      <td>1-kanR-1-0.01</td>\n",
       "      <td>6</td>\n",
       "      <td>180</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.03-145</td>\n",
       "      <td>1-kanR-1-0.03</td>\n",
       "      <td>7</td>\n",
       "      <td>145</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.03-165</td>\n",
       "      <td>1-kanR-1-0.03</td>\n",
       "      <td>7</td>\n",
       "      <td>165</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.03-180</td>\n",
       "      <td>1-kanR-1-0.03</td>\n",
       "      <td>7</td>\n",
       "      <td>180</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.1-145</td>\n",
       "      <td>1-kanR-1-0.1</td>\n",
       "      <td>8</td>\n",
       "      <td>145</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.1-165</td>\n",
       "      <td>1-kanR-1-0.1</td>\n",
       "      <td>8</td>\n",
       "      <td>165</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-0.1-180</td>\n",
       "      <td>1-kanR-1-0.1</td>\n",
       "      <td>8</td>\n",
       "      <td>180</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-1.0-145</td>\n",
       "      <td>1-kanR-1-1.0</td>\n",
       "      <td>9</td>\n",
       "      <td>145</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-1.0-165</td>\n",
       "      <td>1-kanR-1-1.0</td>\n",
       "      <td>9</td>\n",
       "      <td>165</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-kanR-1-1.0-180</td>\n",
       "      <td>1-kanR-1-1.0</td>\n",
       "      <td>9</td>\n",
       "      <td>180</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-0-0.0-80</td>\n",
       "      <td>1-pheS-0-0.0</td>\n",
       "      <td>10</td>\n",
       "      <td>80</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-0-0.0-95</td>\n",
       "      <td>1-pheS-0-0.0</td>\n",
       "      <td>10</td>\n",
       "      <td>95</td>\n",
       "      <td>205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-0-0.0-110</td>\n",
       "      <td>1-pheS-0-0.0</td>\n",
       "      <td>10</td>\n",
       "      <td>110</td>\n",
       "      <td>256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-0-1.0-80</td>\n",
       "      <td>1-pheS-0-1.0</td>\n",
       "      <td>11</td>\n",
       "      <td>80</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-0-1.0-95</td>\n",
       "      <td>1-pheS-0-1.0</td>\n",
       "      <td>11</td>\n",
       "      <td>95</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-0-1.0-110</td>\n",
       "      <td>1-pheS-0-1.0</td>\n",
       "      <td>11</td>\n",
       "      <td>110</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.0-95</td>\n",
       "      <td>1-pheS-1-0.0</td>\n",
       "      <td>12</td>\n",
       "      <td>95</td>\n",
       "      <td>115910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.0-110</td>\n",
       "      <td>1-pheS-1-0.0</td>\n",
       "      <td>12</td>\n",
       "      <td>110</td>\n",
       "      <td>162433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.0-125</td>\n",
       "      <td>1-pheS-1-0.0</td>\n",
       "      <td>12</td>\n",
       "      <td>125</td>\n",
       "      <td>194032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.0001-95</td>\n",
       "      <td>1-pheS-1-0.0001</td>\n",
       "      <td>13</td>\n",
       "      <td>95</td>\n",
       "      <td>115787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.0001-110</td>\n",
       "      <td>1-pheS-1-0.0001</td>\n",
       "      <td>13</td>\n",
       "      <td>110</td>\n",
       "      <td>162034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.0001-125</td>\n",
       "      <td>1-pheS-1-0.0001</td>\n",
       "      <td>13</td>\n",
       "      <td>125</td>\n",
       "      <td>193374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.001-95</td>\n",
       "      <td>1-pheS-1-0.001</td>\n",
       "      <td>14</td>\n",
       "      <td>95</td>\n",
       "      <td>117015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.001-110</td>\n",
       "      <td>1-pheS-1-0.001</td>\n",
       "      <td>14</td>\n",
       "      <td>110</td>\n",
       "      <td>163550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.001-125</td>\n",
       "      <td>1-pheS-1-0.001</td>\n",
       "      <td>14</td>\n",
       "      <td>125</td>\n",
       "      <td>193258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.003-95</td>\n",
       "      <td>1-pheS-1-0.003</td>\n",
       "      <td>15</td>\n",
       "      <td>95</td>\n",
       "      <td>122631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.003-110</td>\n",
       "      <td>1-pheS-1-0.003</td>\n",
       "      <td>15</td>\n",
       "      <td>110</td>\n",
       "      <td>172347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.003-125</td>\n",
       "      <td>1-pheS-1-0.003</td>\n",
       "      <td>15</td>\n",
       "      <td>125</td>\n",
       "      <td>201830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.01-95</td>\n",
       "      <td>1-pheS-1-0.01</td>\n",
       "      <td>16</td>\n",
       "      <td>95</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.01-110</td>\n",
       "      <td>1-pheS-1-0.01</td>\n",
       "      <td>16</td>\n",
       "      <td>110</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.01-125</td>\n",
       "      <td>1-pheS-1-0.01</td>\n",
       "      <td>16</td>\n",
       "      <td>125</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.03-95</td>\n",
       "      <td>1-pheS-1-0.03</td>\n",
       "      <td>17</td>\n",
       "      <td>95</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.03-110</td>\n",
       "      <td>1-pheS-1-0.03</td>\n",
       "      <td>17</td>\n",
       "      <td>110</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.03-125</td>\n",
       "      <td>1-pheS-1-0.03</td>\n",
       "      <td>17</td>\n",
       "      <td>125</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.1-95</td>\n",
       "      <td>1-pheS-1-0.1</td>\n",
       "      <td>18</td>\n",
       "      <td>95</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.1-110</td>\n",
       "      <td>1-pheS-1-0.1</td>\n",
       "      <td>18</td>\n",
       "      <td>110</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-0.1-125</td>\n",
       "      <td>1-pheS-1-0.1</td>\n",
       "      <td>18</td>\n",
       "      <td>125</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-1.0-95</td>\n",
       "      <td>1-pheS-1-1.0</td>\n",
       "      <td>19</td>\n",
       "      <td>95</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-1.0-110</td>\n",
       "      <td>1-pheS-1-1.0</td>\n",
       "      <td>19</td>\n",
       "      <td>110</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V30L/A92Y</th>\n",
       "      <td>V30L/A92Y</td>\n",
       "      <td>6985</td>\n",
       "      <td>1-pheS-1-1.0-125</td>\n",
       "      <td>1-pheS-1-1.0</td>\n",
       "      <td>19</td>\n",
       "      <td>125</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            genotype  genotype_number            condition   base_condition  \\\n",
       "genotype                                                                      \n",
       "V30L/A92Y  V30L/A92Y             6985      1-kanR-0-0.0-80     1-kanR-0-0.0   \n",
       "V30L/A92Y  V30L/A92Y             6985      1-kanR-0-0.0-95     1-kanR-0-0.0   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-kanR-0-0.0-110     1-kanR-0-0.0   \n",
       "V30L/A92Y  V30L/A92Y             6985      1-kanR-0-1.0-80     1-kanR-0-1.0   \n",
       "V30L/A92Y  V30L/A92Y             6985      1-kanR-0-1.0-95     1-kanR-0-1.0   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-kanR-0-1.0-110     1-kanR-0-1.0   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-kanR-1-0.0-160     1-kanR-1-0.0   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-kanR-1-0.0-180     1-kanR-1-0.0   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-kanR-1-0.0-200     1-kanR-1-0.0   \n",
       "V30L/A92Y  V30L/A92Y             6985  1-kanR-1-0.0001-160  1-kanR-1-0.0001   \n",
       "V30L/A92Y  V30L/A92Y             6985  1-kanR-1-0.0001-180  1-kanR-1-0.0001   \n",
       "V30L/A92Y  V30L/A92Y             6985  1-kanR-1-0.0001-200  1-kanR-1-0.0001   \n",
       "V30L/A92Y  V30L/A92Y             6985   1-kanR-1-0.001-160   1-kanR-1-0.001   \n",
       "V30L/A92Y  V30L/A92Y             6985   1-kanR-1-0.001-180   1-kanR-1-0.001   \n",
       "V30L/A92Y  V30L/A92Y             6985   1-kanR-1-0.001-200   1-kanR-1-0.001   \n",
       "V30L/A92Y  V30L/A92Y             6985   1-kanR-1-0.003-160   1-kanR-1-0.003   \n",
       "V30L/A92Y  V30L/A92Y             6985   1-kanR-1-0.003-180   1-kanR-1-0.003   \n",
       "V30L/A92Y  V30L/A92Y             6985   1-kanR-1-0.003-200   1-kanR-1-0.003   \n",
       "V30L/A92Y  V30L/A92Y             6985    1-kanR-1-0.01-145    1-kanR-1-0.01   \n",
       "V30L/A92Y  V30L/A92Y             6985    1-kanR-1-0.01-165    1-kanR-1-0.01   \n",
       "V30L/A92Y  V30L/A92Y             6985    1-kanR-1-0.01-180    1-kanR-1-0.01   \n",
       "V30L/A92Y  V30L/A92Y             6985    1-kanR-1-0.03-145    1-kanR-1-0.03   \n",
       "V30L/A92Y  V30L/A92Y             6985    1-kanR-1-0.03-165    1-kanR-1-0.03   \n",
       "V30L/A92Y  V30L/A92Y             6985    1-kanR-1-0.03-180    1-kanR-1-0.03   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-kanR-1-0.1-145     1-kanR-1-0.1   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-kanR-1-0.1-165     1-kanR-1-0.1   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-kanR-1-0.1-180     1-kanR-1-0.1   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-kanR-1-1.0-145     1-kanR-1-1.0   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-kanR-1-1.0-165     1-kanR-1-1.0   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-kanR-1-1.0-180     1-kanR-1-1.0   \n",
       "V30L/A92Y  V30L/A92Y             6985      1-pheS-0-0.0-80     1-pheS-0-0.0   \n",
       "V30L/A92Y  V30L/A92Y             6985      1-pheS-0-0.0-95     1-pheS-0-0.0   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-pheS-0-0.0-110     1-pheS-0-0.0   \n",
       "V30L/A92Y  V30L/A92Y             6985      1-pheS-0-1.0-80     1-pheS-0-1.0   \n",
       "V30L/A92Y  V30L/A92Y             6985      1-pheS-0-1.0-95     1-pheS-0-1.0   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-pheS-0-1.0-110     1-pheS-0-1.0   \n",
       "V30L/A92Y  V30L/A92Y             6985      1-pheS-1-0.0-95     1-pheS-1-0.0   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-pheS-1-0.0-110     1-pheS-1-0.0   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-pheS-1-0.0-125     1-pheS-1-0.0   \n",
       "V30L/A92Y  V30L/A92Y             6985   1-pheS-1-0.0001-95  1-pheS-1-0.0001   \n",
       "V30L/A92Y  V30L/A92Y             6985  1-pheS-1-0.0001-110  1-pheS-1-0.0001   \n",
       "V30L/A92Y  V30L/A92Y             6985  1-pheS-1-0.0001-125  1-pheS-1-0.0001   \n",
       "V30L/A92Y  V30L/A92Y             6985    1-pheS-1-0.001-95   1-pheS-1-0.001   \n",
       "V30L/A92Y  V30L/A92Y             6985   1-pheS-1-0.001-110   1-pheS-1-0.001   \n",
       "V30L/A92Y  V30L/A92Y             6985   1-pheS-1-0.001-125   1-pheS-1-0.001   \n",
       "V30L/A92Y  V30L/A92Y             6985    1-pheS-1-0.003-95   1-pheS-1-0.003   \n",
       "V30L/A92Y  V30L/A92Y             6985   1-pheS-1-0.003-110   1-pheS-1-0.003   \n",
       "V30L/A92Y  V30L/A92Y             6985   1-pheS-1-0.003-125   1-pheS-1-0.003   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-pheS-1-0.01-95    1-pheS-1-0.01   \n",
       "V30L/A92Y  V30L/A92Y             6985    1-pheS-1-0.01-110    1-pheS-1-0.01   \n",
       "V30L/A92Y  V30L/A92Y             6985    1-pheS-1-0.01-125    1-pheS-1-0.01   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-pheS-1-0.03-95    1-pheS-1-0.03   \n",
       "V30L/A92Y  V30L/A92Y             6985    1-pheS-1-0.03-110    1-pheS-1-0.03   \n",
       "V30L/A92Y  V30L/A92Y             6985    1-pheS-1-0.03-125    1-pheS-1-0.03   \n",
       "V30L/A92Y  V30L/A92Y             6985      1-pheS-1-0.1-95     1-pheS-1-0.1   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-pheS-1-0.1-110     1-pheS-1-0.1   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-pheS-1-0.1-125     1-pheS-1-0.1   \n",
       "V30L/A92Y  V30L/A92Y             6985      1-pheS-1-1.0-95     1-pheS-1-1.0   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-pheS-1-1.0-110     1-pheS-1-1.0   \n",
       "V30L/A92Y  V30L/A92Y             6985     1-pheS-1-1.0-125     1-pheS-1-1.0   \n",
       "\n",
       "           base_condition_number  time  counts  \n",
       "genotype                                        \n",
       "V30L/A92Y                      0    80       1  \n",
       "V30L/A92Y                      0    95       1  \n",
       "V30L/A92Y                      0   110       1  \n",
       "V30L/A92Y                      1    80      12  \n",
       "V30L/A92Y                      1    95       7  \n",
       "V30L/A92Y                      1   110       9  \n",
       "V30L/A92Y                      2   160       0  \n",
       "V30L/A92Y                      2   180       0  \n",
       "V30L/A92Y                      2   200       0  \n",
       "V30L/A92Y                      3   160       1  \n",
       "V30L/A92Y                      3   180       0  \n",
       "V30L/A92Y                      3   200       0  \n",
       "V30L/A92Y                      4   160       0  \n",
       "V30L/A92Y                      4   180       0  \n",
       "V30L/A92Y                      4   200       1  \n",
       "V30L/A92Y                      5   160       0  \n",
       "V30L/A92Y                      5   180       0  \n",
       "V30L/A92Y                      5   200       0  \n",
       "V30L/A92Y                      6   145       1  \n",
       "V30L/A92Y                      6   165       1  \n",
       "V30L/A92Y                      6   180       1  \n",
       "V30L/A92Y                      7   145       2  \n",
       "V30L/A92Y                      7   165       1  \n",
       "V30L/A92Y                      7   180       1  \n",
       "V30L/A92Y                      8   145       3  \n",
       "V30L/A92Y                      8   165       3  \n",
       "V30L/A92Y                      8   180       1  \n",
       "V30L/A92Y                      9   145       1  \n",
       "V30L/A92Y                      9   165       2  \n",
       "V30L/A92Y                      9   180       2  \n",
       "V30L/A92Y                     10    80     155  \n",
       "V30L/A92Y                     10    95     205  \n",
       "V30L/A92Y                     10   110     256  \n",
       "V30L/A92Y                     11    80      24  \n",
       "V30L/A92Y                     11    95      23  \n",
       "V30L/A92Y                     11   110      21  \n",
       "V30L/A92Y                     12    95  115910  \n",
       "V30L/A92Y                     12   110  162433  \n",
       "V30L/A92Y                     12   125  194032  \n",
       "V30L/A92Y                     13    95  115787  \n",
       "V30L/A92Y                     13   110  162034  \n",
       "V30L/A92Y                     13   125  193374  \n",
       "V30L/A92Y                     14    95  117015  \n",
       "V30L/A92Y                     14   110  163550  \n",
       "V30L/A92Y                     14   125  193258  \n",
       "V30L/A92Y                     15    95  122631  \n",
       "V30L/A92Y                     15   110  172347  \n",
       "V30L/A92Y                     15   125  201830  \n",
       "V30L/A92Y                     16    95      36  \n",
       "V30L/A92Y                     16   110      67  \n",
       "V30L/A92Y                     16   125      83  \n",
       "V30L/A92Y                     17    95      49  \n",
       "V30L/A92Y                     17   110      67  \n",
       "V30L/A92Y                     17   125      70  \n",
       "V30L/A92Y                     18    95      53  \n",
       "V30L/A92Y                     18   110      61  \n",
       "V30L/A92Y                     18   125      53  \n",
       "V30L/A92Y                     19    95      39  \n",
       "V30L/A92Y                     19   110      47  \n",
       "V30L/A92Y                     19   125      68  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# counter = 0\n",
    "# base_condition_number = {}\n",
    "# for c in condition_df_with_time.index:\n",
    "#     base_condition = \"-\".join(c.split(\"-\")[:-1])\n",
    "#     if base_condition not in base_condition_number:\n",
    "#         base_condition_number[base_condition] = counter\n",
    "#         counter += 1\n",
    "        \n",
    "# base_condition_number\n",
    "#count_df.loc[count_df[\"genotype\"] == \"wt\",:] #.sort_values([\"genotype_number\",\"base_condition_number\",\"time\"])\n",
    "#count_df.index = np.arange(len(count_df[\"genotype\"]),dtype=int)\n",
    "count_df.loc[count_df[\"counts\"] == 201830,:]\n",
    "count_df.loc[count_df[\"genotype\"] == \"V30L/A92Y\",:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee2902e8-b1d4-4888-900a-470b58a4c52b",
   "metadata": {},
   "outputs": [],
   "source": [
    "g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b12c7342-4608-4433-9382-4c62dedf5947",
   "metadata": {},
   "outputs": [],
   "source": [
    "ct = condition_df_with_time.copy()\n",
    "ct[\"cfu_per_mL\"] = 0.0\n",
    "\n",
    "all_pops = {}\n",
    "for i in range(ln_pop_array.shape[1]):\n",
    "    \n",
    "    c = condition_df.loc[condition_df.index[i],[\"replicate\",\n",
    "                                                \"marker\",\n",
    "                                                \"select\",\n",
    "                                                \"iptg\"]]\n",
    "    mask = (ct[\"replicate\"] == c[\"replicate\"]) & \\\n",
    "           (ct[\"marker\"] == c[\"marker\"]) & \\\n",
    "           (ct[\"select\"] == c[\"select\"]) & \\\n",
    "           (ct[\"iptg\"] == c[\"iptg\"])\n",
    "\n",
    "    delta_time = np.array(ct.loc[mask,\"time\"],dtype=float)\n",
    "    delta_time[1:] = delta_time[1:] - delta_time[:-1]\n",
    "\n",
    "    condition_ids = ct.index[mask]\n",
    "\n",
    "    p = None\n",
    "    for j in range(len(condition_ids)):\n",
    "        \n",
    "        if p is None:\n",
    "            p = ln_pop_array[:,i]\n",
    "        p = tfscreen.grow_for_time(ln_pop_array=p,\n",
    "                                   growth_rates=bact_condition_k[:,i],\n",
    "                                   t=dt)\n",
    "        all_pops[condition_ids[j]] = p\n",
    "\n",
    "        ct.loc[condition_ids[j],\"cfu_per_mL\"] = np.sum(np.exp(p))\n",
    "\n",
    "num_samples = len(ct)\n",
    "ct[\"num_reads\"] = int(total_num_reads//num_samples)\n",
    "bacteria_as_int = np.arange(len(bacteria),dtype=int)\n",
    "\n",
    "sequencing_dataframes = []\n",
    "for condition in tqdm(all_pops):\n",
    "\n",
    "    num_reads = ct.loc[condition,\"num_reads\"]\n",
    "    \n",
    "    pop = all_pops[condition]\n",
    "    w = np.exp(pop)\n",
    "    w = w/np.sum(w)\n",
    "    sample = np.random.choice(bacteria_as_int,size=num_reads,p=w)\n",
    "\n",
    "    this_df = genotype_df.copy()\n",
    "    this_df.insert(loc=0,column=\"condition\",value=condition)\n",
    "    this_df[\"counts\"] = 0\n",
    "\n",
    "    seen, counts = np.unique(sample,return_counts=True)\n",
    "    bacteria_seen = bacteria[seen]\n",
    "\n",
    "    if not issubclass(type(bacteria_seen[0]),str):\n",
    "        \n",
    "        new_bacteria_seen = []\n",
    "        new_counts = []\n",
    "        \n",
    "        for i in range(len(bacteria_seen)):\n",
    "            genotypes = bacteria_seen[0]\n",
    "            sub_counts = counts[0]\n",
    "            plasmids_pulled = np.random.choice(genotypes,size=sub_counts)\n",
    "            this_g, this_count = np.unique(plasmids_pulled,return_counts=True)\n",
    "            new_bacteria_seen.extend(this_g)\n",
    "            new_counts.extend(this_count)\n",
    "\n",
    "        bacteria_seen = new_bacteria_seen\n",
    "        counts = new_counts\n",
    "    \n",
    "    this_df.loc[bacteria_seen,\"counts\"] = counts\n",
    "\n",
    "    sequencing_dataframes.append(this_df)\n",
    "    \n",
    "count_df = pd.concat(sequencing_dataframes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf102f96-bfa9-411f-a065-13384fad01ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "count_df[count_df[\"counts\"] > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f68aa88-4360-4b53-be4d-94e95aab2407",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# Get the population over time for each each selection and iptg condition\n",
    "pops_vs_time = growth_with_selection(ln_pop_array,\n",
    "                                     bact_condition_k,\n",
    "                                     time_points=sample_times)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3579b374-8aef-47f1-a425-54f257552356",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# -----------------------------------------------------------------------------\n",
    "# Sequence samples and summarize results\n",
    "df, genotype_df, condition_df = sequence_and_collate(pops_vs_time=pops_vs_time,\n",
    "                                                     iptg_concs=iptg_concs,\n",
    "                                                     sample_times=sample_times,\n",
    "                                                     input_library=input_library,\n",
    "                                                     all_genotypes=all_genotypes,\n",
    "                                                     num_reads_per_condition=num_reads_per_condition)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3316127f-0f26-4ab4-ab9b-cd8ee47a6f81",
   "metadata": {},
   "outputs": [],
   "source": [
    "ct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "075de7f2-b4ee-4db4-9e05-1f6febb351b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "print(\"Generating initial populations\",flush=True)\n",
    "if input_library.ndim == 1:\n",
    "\n",
    "    bacteria, ln_pop_array = tfscreen.thaw_glycerol_stock(input_library)\n",
    "    bacteria, counts = np.unique(bacteria,return_counts=True)\n",
    "    ln_pop_array = np.log(np.exp(ln_pop_array[0])*counts)\n",
    "\n",
    "else:\n",
    "\n",
    "    to_thaw = []\n",
    "    indexer = {}\n",
    "    for i in tqdm(range(input_library.shape[0])):\n",
    "        \n",
    "        tmp_genos = list(input_library[i][input_library[i] != \"-\"])\n",
    "        tmp_genos.sort()\n",
    "        tmp_genos = tuple(tmp_genos)\n",
    "        \n",
    "        if tmp_genos not in indexer:\n",
    "            indexer[tmp_genos] = i\n",
    "        \n",
    "        to_thaw.append(indexer[tmp_genos])\n",
    "\n",
    "    to_thaw = np.array(to_thaw)\n",
    "    reverse_indexer = dict(zip(indexer.values(),indexer.keys()))\n",
    "\n",
    "    bacteria, ln_pop_array = tfscreen.thaw_glycerol_stock(to_thaw)\n",
    "    bacteria, counts = np.unique(bacteria,return_counts=True)\n",
    "    ln_pop_array = np.log(np.exp(ln_pop_array[0])*counts)\n",
    "\n",
    "    bacteria = np.array([reverse_indexer[b] for b in bacteria],dtype=object)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84b44e85-d7e2-4cd3-98b2-a4e6b4939b46",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(ln_pop_array,bins=np.arange(-6,6,0.1))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d47cfd4f-2e23-4524-a043-b61beed42e7a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05da2984-f47d-4edd-902a-7f734aa6900f",
   "metadata": {},
   "outputs": [],
   "source": [
    "bacteria, population = tfscreen.thaw_glycerol_stock(out)\n",
    "bacteria, counts = np.unique(bacteria,return_counts=True)\n",
    "population = np.log(np.exp(population[0])*counts)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d294b4ae-e594-4a12-8bb2-edb082766d2e",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "genotype_to_idx = dict(zip(list(genotype_df.index),\n",
    "                           range(len(genotype_df.index))))\n",
    "\n",
    "num_conditions = condition_df.shape[0]\n",
    "num_bacteria = len(input_library)\n",
    "\n",
    "condition_k = np.array(phenotype_df[\"growth_rate\"])\n",
    "condition_k = np.reshape(condition_k,\n",
    "                         (len(condition_k)//num_conditions,\n",
    "                          num_conditions))\n",
    "base_delta_k = np.array(genotype_df[\"growth_rate_effect\"])\n",
    "\n",
    "\n",
    "if input_library.ndim == 1:\n",
    "    \n",
    "    idx = np.array([genotype_to_idx[g] for g in input_library])\n",
    "\n",
    "    clone_genotypes = input_library\n",
    "    clone_base_delta_k = base_delta_k[idx]\n",
    "    clone_condition_k = condition_k[idx]\n",
    "\n",
    "else:\n",
    "\n",
    "    clone_genotypes = []\n",
    "    clone_condition_k = np.zeros((num_bacteria,num_conditions),dtype=float)\n",
    "    clone_base_delta_k = np.zeros(num_bacteria)\n",
    "    \n",
    "    for i in tqdm(range(len(input_library))):\n",
    "    \n",
    "        genotypes = input_library[i][input_library[i] != \"-\"]\n",
    "        idx = np.array([genotype_to_idx[g] for g in genotypes])\n",
    "    \n",
    "        clone_genotypes.append(genotypes)\n",
    "        clone_condition_k[i] = np.mean(condition_k[idx,:],axis=0)\n",
    "        clone_base_delta_k[i] = np.mean(base_delta_k[idx])\n",
    "    \n",
    "    clone_genotypes = list(clone_genotypes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2fb79d4-809c-405c-9b8f-efcbf3aea047",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_library = [\n",
    "x, y = np.unique(input_library,return_counts=True)\n",
    "input_library.shape\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a6d978a-7ce2-491e-9de5-f5c12ff6e509",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_k[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "880e5699-0403-4a5c-807d-ada486ae7f89",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"obs.csv\")\n",
    "genotype_df.to_csv(\"genotype.csv\")\n",
    "cfu_df.to_csv(\"cfu.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "583db7fe-ba65-4817-92d1-51fbf35a7cab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# libraries --> dictionary keying each library to a list of mutations in each clone\n",
    "# \n",
    "# lib_phenotypes\n",
    "# [{'clone': ['V30N'],\n",
    "#    'ddG': array([2.0162, 1.8317, 1.8317, 2.0162, 2.0162, 1.8317, 1.8317, 2.0162,\n",
    "#           1.8317, 1.8312, 1.8312, 1.8312, 1.831 , 1.8309, 1.8309, 1.8309,\n",
    "#           0.    ]),\n",
    "#    'fx_occupied': array([0.0007608 , 0.00103465, 0.00426162, 0.01198636, 0.02997511,\n",
    "#           0.04687013, 0.05698534, 0.06055376]),\n",
    "#    'fx_folded': array([1.        , 1.        , 0.99999996, 0.99999979, 0.99999928,\n",
    "#           0.99999875, 0.99999842, 0.9999983 ]),\n",
    "#    'obs': array([0.0007608 , 0.00103465, 0.00426162, 0.01198636, 0.02997509,\n",
    "#           0.04687007, 0.05698525, 0.06055365]),\n",
    "#    'base_growth_rate': np.float64(0.005445439477620463),\n",
    "#    'sel_kan': array([3.98258294e-05, 5.41413358e-05, 2.22833015e-04, 6.26648000e-04,\n",
    "#           1.56701933e-03, 2.45021391e-03, 2.97899038e-03, 3.16553093e-03]),\n",
    "#    'kan_p0g': np.float64(5.4454394776204634e-08),\n",
    "#    'kan_p1g': np.float64(0.004356351582096371),\n",
    "#    'sel_pheS': array([0.00431658, 0.00430226, 0.00413357, 0.00372976, 0.00278939,\n",
    "#           0.00190619, 0.00137742, 0.00119088]),\n",
    "#    'pheS_p0g': np.float64(0.004356351582096371),\n",
    "#    'pheS_p1g': np.float64(5.4454394776204634e-08)},\n",
    "\n",
    "# input_library and all_genotypes\n",
    "# input_library -> array([214305,   8702, 169508, ...,  81919, 281807,  70355],shape=(10000000,))\n",
    "\n",
    "# all_genotypes -> list with full genotype information (like lib_phenotypes) indexed by input library.\n",
    "# ln_pop_array : log population of each clone in initial library before selection (indexed by input_library)\n",
    "# base_growth_rates : base growth rate of each clone (indexed by input_library)\n",
    "# growth_rates : dictionary keyed to selection. values are N x num_iptg arrays where N is indexed by input_library\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bfc50dd-ea1a-4d8a-9900-7f70b3f5d8ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "all_genotypes = []\n",
    "for lib in libraries:\n",
    "    all_genotypes.extend([\"/\".join(g) for g in libraries[lib]])\n",
    "\n",
    "genotype_df = build_genotype_df(all_genotypes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b423ac9d-9bb5-4534-8f5a-37f1cdf975ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "libraries[('1',)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9c7a7ff-9845-4159-9243-fe96b6142d91",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
